flink-clients/src/main/java/org/apache/flink/client/program/StreamContextEnvironment.java

```diff
@@ -169,12 +242,42 @@ public static void setAsContext(
                             mergedConfiguration,
                             userCodeClassLoader,
                             enforceSingleJobExecution,
-                            suppressSysout);
+                            suppressSysout,
+                            allowConfigurations,
+                            errors);
                 };
         initializeContextEnvironment(factory);
     }
 
     public static void unsetAsContext() {
         resetContextEnvironment();
     }
+
+    private List<String> collectNotAllowedConfigurations() {
```

Comment:
I'm not sure about the approach with the configuration diffing here 🤔

Would it make sense to simply make the configurations objects exposed to the user immutable? (something along the lines of `java.util.Collections#unmodifiableCollection`)?

Reply:
I do not think we can do this because it would break the behavior of users modifying the configurations. In general, I agree 100% the current situation is definitely not good that users can mutate these objects and we have to handle it now.

Reply:
Wouldn't it actually make more sense? We simply don't want user to be able to mutate any configuration.

The only difference would be failing earlier and providing user with a full stack trace of the problematic call.

Reply:
There could be also a half-way option of letting user "mutate" the configuration as long as it doesn't change the default value.

Reply:
I like the idea but I think the effort will be significant because we need to monkey patch all methods in the CheckpointConfig[1] and ExecutionConfig[2]. 
We deemed it as less maintainable.

[1] https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/CheckpointConfig.java
[2] https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java

Reply:
I tried to rework the configuration management a bit, unfortunately, did not succeed. The problem is that some configurations you can set at the `ExecutionConfig` or `CheckpointConfig` do not have direct `ConfigOption`. All the configurations basically load a user class like some custom serializer.

We would need to consolidate the resolution of these classes first before unifying that all configurations are reflected through `ConfigOption` but I do not see this PR as a good point to rework the configuration management.

Reply:
Agreed, until we get rid of Execution and Checkpoint config, this would be tricky to achieve.